### DISCUSSION TITLE
SSR: does rsbuild need to load the SSR bundle on every request?

### DISCUSSION THREAD

**Discussion by jdavis** [web-infra-dev/rsbuild#4038] [General]
Hi,

I'm looking over the docs and in the [SSR section](https://rsbuild.dev/guide/advanced/ssr#custom-server), I see this:
```js
const serverRender = (serverAPI) => async (_req, res) => {
  // Load SSR bundle
  const indexModule = await serverAPI.environments.ssr.loadBundle('index');

  const markup = indexModule.render();

  const template = await serverAPI.environments.web.getTransformedHtml('index');

  // Insert SSR rendering content into HTML template
  const html = template.replace('<!--app-content-->', markup);

  res.writeHead(200, {
    'Content-Type': 'text/html',
  });
  res.end(html);
};
```

specifically this line is called on every request: `const indexModule = await serverAPI.environments.ssr.loadBundle('index');`

Is that a requirement that it be loaded on every request? Or can it be loaded and reused for performance gains? The thing it seems like you'd lose on reusing the already loaded bundle would be if it is updated and rebuilt but that could be solved in another way.

**Comment by 9aoy** [comment]
Thanks for your reminder. It seems that adding caching capabilities to the `loadBundle` when the file has not changed could indeed be beneficial.

**Reply by 9aoy** [reply]
Hi, rsbuild [v1.1.9](https://github.com/web-infra-dev/rsbuild/releases/tag/v1.1.9) already supports environment API result caching, hope it will be helpful to you.

**Reply by jdavis** [reply]
Thank you so much for the feature and the quick turnaround!
